### PR TITLE
fenicsx: Flux BC support

### PR DESCRIPTION
**Pull request by jhdark**
## Proposed changes

With this change, users will be able to impose flux boundary conditions:

$-D\nabla c_{\mathrm{m}} \cdot \mathbf{n} = f(x, y, z, t) \ \mathrm{on} \ \partial\Omega$

Users can implement this like:

```python3
my_model.boundary_conditions = [
        F.FluxBC(subdomain=right, value=10, species=H),
]
```

## Types of changes

What types of changes does your code introduce to FESTIM?

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code refactoring
- [ ] Documentation Update (if none of the other choices apply)
- [ ] New tests

## Checklist

- [x] Black formatted
- [ ] Unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: festim/boundary_conditions/flux_bc.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+        if not isinstance(value, (fem.Function, fem.Constant, np.ndarray)):
+            raise TypeError(
+                f"Value must be a dolfinx.fem.Function, dolfinx.fem.Constant, or a np.ndarray not {type(value)}"
+            )
+        self._value_fenics = value
+
+    @property
+    def time_dependent(self):
+        if self.value is None:
+            return False
```
Comment (RemDelaporteMathurin) [inline festim/boundary_conditions/flux_bc.py:66]: Isn't this a bit dangerous? Say I check if it's time-dependent before assigning it a value by mistake, I may think it's not time-dpeendent but then it becomes time dependent if the value is say a function of time.

Should we raise an error instead?

File: festim/boundary_conditions/flux_bc.py
Comment (RemDelaporteMathurin) [inline festim/boundary_conditions/flux_bc.py:1]: Should we start off by having `FluxBCBase` and `ParticleFluxBC` and an empty `HeatFluxBC` class?

File: festim/boundary_conditions/flux_bc.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+                )
+                self.value_fenics.interpolate(self.bc_expr)
+
+    def update(self, t):
+        """Updates the source value
+
+        Args:
+            t (float): the time
+        """
+        if callable(self.value):
```
Comment (RemDelaporteMathurin) [inline festim/boundary_conditions/flux_bc.py:144]: Is this needed? I would replace by `if self.time_dependent`? what if it isn't? should we raise an error? or just a print?

**Review by RemDelaporteMathurin — CHANGES_REQUESTED** [review]

File: festim/boundary_conditions/flux_bc.py
Comment (jhdark) [reply festim/boundary_conditions/flux_bc.py:1]: For sure!

File: festim/boundary_conditions/flux_bc.py (hunk 2 above)
Comment (jhdark) [reply festim/boundary_conditions/flux_bc.py]: This is how we also handle it in sources, and Dirichlet bc, so maybe we could change this for all of them in another PR?

File: festim/boundary_conditions/flux_bc.py (hunk 1 above)
Comment (jhdark) [reply festim/boundary_conditions/flux_bc.py:55]: Yeah this is valid, but would be worth changing the other cases of this property in other classes too

File: festim/boundary_conditions/flux_bc.py [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+        >>> ParticleFluxBC(subdomain=my_subdomain, value=lambda T: 1 + T, species="H")
+        >>> ParticleFluxBC(subdomain=my_subdomain, value=lambda x, t: 1 + x[0] + t, species="H")
+    """
+
+    def __init__(self, subdomain, value, species):
+        super().__init__(subdomain=subdomain, value=value)
+        self.species = species
+
+        self.value_fenics = None
+        self.bc_expr = None
```
Comment (RemDelaporteMathurin) [inline festim/boundary_conditions/flux_bc.py:113]: can be removed

File: festim/boundary_conditions/flux_bc.py [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+                # store the expression of the boundary condition
+                # to update the value_fenics later
+                self.bc_expr = fem.Expression(
+                    self.value(**kwargs),
+                    function_space.element.interpolation_points(),
+                )
+                self.value_fenics.interpolate(self.bc_expr)
+
+
+class HeatFluxBC(FluxBCBase):
```
Comment (RemDelaporteMathurin) [inline festim/boundary_conditions/flux_bc.py:179]: With this implementation, can the HeatFluxBC be temperature-dependent?
I don't think so. Users should be able to do:

```python
my_flux = HeatFluxBc(lambda T: h * (T-T_ext))
```

Comment (RemDelaporteMathurin) [reply festim/boundary_conditions/flux_bc.py:179]: With this change, I think that a lot of the code will be duplicated from `ParticleFlux` therefore all this functionality can be moved to `FluxBCBase`

File: festim/heat_transfer_problem.py [hunk 5]
```diff
@@ -53,7 +53,9 @@ def boundary_conditions(self):
 
     @boundary_conditions.setter
     def boundary_conditions(self, value):
-        if not all(isinstance(bc, F.FixedTemperatureBC) for bc in value):
+        if not all(
+            isinstance(bc, (F.FixedTemperatureBC, F.HeatFluxBC)) for bc in value
+        ):
             raise TypeError(
                 "boundary_conditions must be a list of festim.FixedTemperatureBC objects"
```
Comment (RemDelaporteMathurin) [inline festim/heat_transfer_problem.py:60]:
```suggestion
                "boundary_conditions must be a list of festim.FixedTemperatureBC or festim.HeatFluxBC objects"
```

File: festim/heat_transfer_problem.py
Comment (RemDelaporteMathurin) [inline festim/heat_transfer_problem.py:1]: In this, shouldn't we make sure we update the value of a time-dependent HeatFluxBC? Is it already done?

**Review by RemDelaporteMathurin — CHANGES_REQUESTED** [review]

File: festim/boundary_conditions/flux_bc.py (hunk 4 above)
Comment (jhdark) [reply festim/boundary_conditions/flux_bc.py:179]: Ah of course yes

File: festim/boundary_conditions/flux_bc.py (hunk 2 above)
Comment (RemDelaporteMathurin) [reply festim/boundary_conditions/flux_bc.py:133]: Sounds good with me

File: festim/boundary_conditions/flux_bc.py (hunk 4 above)
Comment (RemDelaporteMathurin) [reply festim/boundary_conditions/flux_bc.py:176]: oh wow that's a lot of code that went to FluxBCBase haha

File: festim/heat_transfer_problem.py
Comment (RemDelaporteMathurin) [reply festim/heat_transfer_problem.py:1]: https://github.com/festim-dev/FESTIM/blob/0375d58c0267aa5c56f6d82b5e97aa8af6a9e642/festim/heat_transfer_problem.py#L337-L345

yes it is arleady done here

File: festim/boundary_conditions/flux_bc.py (hunk 2 above)
Comment (RemDelaporteMathurin) [reply festim/boundary_conditions/flux_bc.py:133]: I opened an issue #743 to remind us of this

File: festim/boundary_conditions/flux_bc.py (hunk 1 above)
Comment (RemDelaporteMathurin) [reply festim/boundary_conditions/flux_bc.py]: Opened an issue to track progress on this #744

**Review by RemDelaporteMathurin — APPROVED** [review]

**Pull request merged**